### PR TITLE
:book: Add IPAM introduction document

### PR DESCRIPTION
**Pull request by ren-yifei**
Add IPAM introduction document

Fixes: #228

### CODE REVIEW

**Comment by smoshiur1237** [comment]
Please add your md file reference to SUMMARY.md

**Review by fmuyassarov — COMMENTED** [review]
Thanks @ren-yifei for the document. Can we perhaps talk a bit more what is IPAM, what it aims to solve. An example document I found is: http://metal3.io/blog/2020/07/06/IP_address_manager.html

**Comment by ren-yifei** [comment]
I updated the IPAM introduction part with the content mentioned by @fmuyassarov . Please check again.

File: docs/user-guide/src/ipam/ipam.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+recreate the status in the target cluster without losing information. All
+objects must, through a tree of owner references, be attached to the cluster
+object, for the pivoting to proceed properly.
+
+Moreover, there are use cases
+that the users want to specify multiple non-continuous ranges of IP addresses,
+use the same pool across multiple Template objects, or rule out some IP
+addresses that might be in use for any reason after the deployment.
+
+The IPAM is introduced to manages the allocations of IP subnets according to
```
Comment (fmuyassarov) [inline docs/user-guide/src/ipam/ipam.md:28]:
```suggestion
The IPAM is introduced to manage the allocations of IP subnets according to
```

File: docs/user-guide/src/ipam/ipam.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+the requests without handling any use of those addresses. The IPAM adds the
+flexibility by providing the address right before provisioning the node. It can
+share a pool across machine deployments or KCP, allow non-continuous pools
+and external IP management by using IPAddress CRs, offer predictable IP
```
Comment (fmuyassarov) [inline docs/user-guide/src/ipam/ipam.md:32]:
```suggestion
share a pool across MachineDeployments (MD) or KCP, allow non-continuous pools
and external IP management by using IPAddress CRs, offer predictable IP
```

File: docs/user-guide/src/ipam/ipam.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+* **IPAddress**: an IP address allocation
+
+## Deployment
+
+You can deploy IPAM using `make deploy` or deploy an example pool using
+`make deploy-examples`. You can also run IPAM controller locally using
+`make run`.
+
+For more information about this controller and related repositories, see
+[metal3.io](http://metal3.io/).
```
Comment (fmuyassarov) [inline docs/user-guide/src/ipam/ipam.md:52]: I feel this part can be a bit extended, please have a look also to document at https://github.com/metal3-io/ip-address-manager/blob/main/docs/deployment_workflow.md

File: docs/user-guide/src/ipam/ipam.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+The IPAM project provides a controller to manage static IP address allocations
+in [Cluster API Provider Metal3](https://github.com/metal3-io/cluster-api-provider-metal3/).
+
+In CAMP3, the Network Data need to be passed to ironic through the BMH (Bare
+Metal Host). CAPI addresses the deployment of Kubernetes clusters and nodes, using the
+Kubernetes API. As such, it uses objects such as Machine Deployments (similar
+to deployments for pods) that takes care of creating the requested number of
+machines, based on templates. The replicas can be increased by the user,
+triggering the creation of new machines based on the provided templates.
```
Comment (fmuyassarov) [inline docs/user-guide/src/ipam/ipam.md:11]: This part seems not directly related to the context, I would rather drop these lines and extend lines 12-14. You have mentioned that it is not possible to provide static IP addresses, but can we explain why?

File: docs/user-guide/src/ipam/ipam.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+Considering the Kubeadm Control Plane and machine deployment features in
+Cluster API, it is not possible to provide static IP addresses
+for each machine before the actual deployments.
+
+In addition, all the resources from the source cluster must support the CAPI
+pivoting, i.e. being copied and recreated in the target cluster. This means
+that all objects must contain all needed information in their spec field to
+recreate the status in the target cluster without losing information. All
+objects must, through a tree of owner references, be attached to the cluster
+object, for the pivoting to proceed properly.
```
Comment (fmuyassarov) [inline docs/user-guide/src/ipam/ipam.md:21]: Here also I am not sure what's the relationship of this context TBH

**Review by fmuyassarov — COMMENTED** [review]
Thanks @ren-yifei . I have left few more comments

File: docs/user-guide/src/ipam/ipam.md (hunk 1 above)
Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: Fixed

File: docs/user-guide/src/ipam/ipam.md (hunk 2 above)
Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: Fixed in this line and also the above lines.

File: docs/user-guide/src/ipam/ipam.md (hunk 3 above)
Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: Extended, and added the deletion part.

File: docs/user-guide/src/ipam/ipam.md (hunk 5 above)
Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: Removed.

File: docs/user-guide/src/ipam/ipam.md (hunk 4 above)
Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: These lines are removed. However, I have problems describing the reason why the static IP addresses are not possible to be provided.

Comment (ren-yifei) [reply docs/user-guide/src/ipam/ipam.md]: Updated, please check.

File: docs/user-guide/src/SUMMARY.md [hunk 6]
```diff
@@ -24,5 +24,6 @@
         - [Node Reuse]()
         - [Pivoting]()
 - [Ip-address-manager]()
+    - [Introduction](ipam/ipam.md)
```
Comment (fmuyassarov) [inline docs/user-guide/src/SUMMARY.md:27]: Can you please move this document under `[Ip-address-manager]()`, I guess we don't need a  new section

**Review by fmuyassarov — COMMENTED** [review]
/lgtm

**Comment by fmuyassarov** [comment]
Since @ren-yifei has left the Metal3, I have done the rebase for her.
/hold cancel
/lgtm

**Pull request merged**